rfcs/0000-graphql-api.md

```diff
@@ -0,0 +1,433 @@
+## Summary
+[summary]: #summary
+
+A GraphQL api design to allow the wallet app (and future similar applications) to communicate with a full node.
+
+## Motivation
+
+[motivation]: #motivation
+
+Currently, the main use case for this is to support the development of a wallet application which can
+communicate with a full node which is running either remotely or locally in another process.
+
+Related Goals:
+- Make app development using the api relatively easy and fast.
+- Make it easier for people to communicate programmatically with a full node.
+- Consumers of the api should not be required to write in a specific language.
+- Eventually be able to rewrite the command-line client to use the same api.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+This design is heavily influenced by which data is stored on the node and which is intended to be stored by the client.
+Most importantly, we assume the node knows about some private keys which it associates to public keys for which it will
+try to store related transactions indefinitely (where "related" means being sent to or from the public key). We may also
+need to store information about blocks that these keys have created.
+We will likely want to at least optionally cache the history of the blocks/payments on the client side as well,
+just to avoid needing to query it all every time.
+
+Note: There are several custom scalars defined at the top for readability. These will be `String`s in the final implementation due to the complications involved in the encoding of custom scalars not being expressed in the schema and needing to be implemented symmetrically on the client and server.
+
+Note: Public keys will all be the "compressed" public keys that are used elsewhere in the node.
+
+```graphql
+# Note: this will all be strings in the actual api.
+scalar Date
+scalar PublicKey
+scalar PrivateKey
+scalar UInt64
+scalar UInt32
+
+enum ConsensusStatus {
+  SUBMITTED
+  INCLUDED # Included in any block
+  FINALIZED
+  SNARKED
+  FAILED
+}
+
+type ConsensusState {
+  status: ConsensusStatus!
+  estimatedPercentConfirmed: Float!
+}
+
+type Payment {
+  nonce: Int!,
+  submittedAt: Date!
+  includedAt: Date
+  from: PublicKey! 
+  to: PublicKey!
+  amount: UInt64!
+  fee: UInt32!,
+  memo: String
+}
+
+type PaymentUpdate {
+  payment: Payment
+  consensus: ConsensusState!
+}
+
+enum SyncStatus {
+  ERROR
+  BOOTSTRAP # Resyncing
+  STALE # You haven't seen any activity recently
+  SYNCED
+}
+
+type SyncUpdate {
+  status: SyncStatus!
+  estimatedPercentSynced: Float!
+  description: String
+}
+
+type SnarkWorker {
+  key: PublicKey!
+  fee: UInt32!
+}
+
+type SnarkFee {
+  snarkCreator: PublicKey!
+  fee: UInt32!
+}
+
+type SnarkFeeUpdate {
+  fee: SnarkFee
+  consensus: ConsensusState!
+}
+
+type Block {
+  coinbase: UInt32!
+  creator: PublicKey!
+  payments: [Payment]!
+  snarkFees: [SnarkFee]!
+}
+
+type BlockUpdate {
+  block: Block!
+  consensus: ConsensusState!
+}
+
+type Balance {
```

Comment:
We should revisit the right way to communicate total/unknown balance as semantically others' wallets won't have total/unknown only a single number since we won't be tracking arbitrary transactions from others' wallets.